### DISCUSSION TITLE
HW UART

### DISCUSSION THREAD

**Discussion by msasdk** [UtilitechAS/amsreader-firmware#684] [Ideas]
To my knowledge, on all ESP32 devices, all UART channels can map to any GPIO pin, that is HW UART is available on any GPIO so why not provide that option? One idea would be to change the UI to something like this:

|  Channel | pin   |
|---|---|
| UART0  | 9  |

From C3 datasheet:
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/13538204/e09fbf5b-e199-40a0-8238-ccc5ce143773)

**Comment by ArnieO** [comment]
Sorry - I'm not sure I understand what you mean.

I also have not tried the firmware with the C3, but I assume you during initial setup choose "Generic ESP32-C3 module" or something like that. You should then see on your control page fields where you can select which GPIOs are used for the various functions used by the firmware. 

Are you asking for the flexibility to remap UART0 to another GPIO? If so, I'm not sure if that is within the scope of this firmware.
You will at least need to argue a bit harder! 😁

**Comment by gskjold** [comment]
On S2 and C3, uart1 is remapped to selected GPIO. Is this also possible on vanilla esp32? I was under the understanding that it was not possible there

**Reply by ArnieO** [reply]
I think you're right. The newest ESP32s are flexible, not the older ones. 
So this could become complex to code (and with risk of errors). 
I do not recommend we go there.

**Reply by msasdk** [reply]
This is from the classic ESP32 datasheet, it's as flexible as the others. 
![image](https://github.com/UtilitechAS/amsreader-firmware/assets/13538204/b70c89af-76a8-4a03-a973-f40024528255)
@gskjold Are you saying that on S2 and C3, HW uart(1) is already used when when you choose a gpio pin, and not soft uart?

**Reply by gskjold** [reply]
> Are you saying that on S2 and C3, HW uart(1) is already used when when you choose a gpio pin, and not soft uart?

Yes. Will look into change for classic ESP32

**Reply by msasdk** [reply]
so perhaps the UI should indicate this, or at least the wiki
https://github.com/UtilitechAS/amsreader-firmware/wiki/GPIO-configuration#han

- UART - These are available hardware serial ports. It is recommended to use hardware serial for stability

- GPIO - General pin on your ESP. **Software serial will be set up for this pin**

**Comment by gskjold** [comment]
This have changed for v2.3, we are now always using hw serial for esp32.
